panel/io/compile.py

```diff
@@ -0,0 +1,177 @@
+        esm_code = esm_code.replace(replace, package_name)
+
+    return esm_code, packages
+
+
+def extract_dependencies(component: type[ReactiveESM]) -> tuple[str, dict[str, any]]:
+    """
+    Extracts dependencies from a ReactiveESM component by parsing its
+    importmap and the associated code and replaces URL import
+    specifiers with package imports.
+
+    Arguments
+    ---------
+    component: type[ReactiveESM]
+        The ReactiveESM component to extract a dependency definition from.
+
+    Returns
+    -------
+    code: str
+        Code where the URL imports have been replaced by package imports.
+    dependencies: dict[str, str]
+        A dictionary of package dependencies and their versions.
+    """
+    importmap = component._process_importmap()
+    if not importmap:
+        raise ValueError(
+            f'{component.__name__} did not declare an importmap '
+            'and cannot be compiled.'
+        )
+
+    # Extract dependencies and versions
+    dependencies = {}
+    for key, url in importmap["imports"].items():
+        match = re.search(r'@(\d+\.\d+\.\d+)', url)
+        if key.endswith('/'):
+            key = key[:-1]
+
+        if match:
+            dependencies[key] = f"^{match.group(1)}"
+        else:
+            dependencies[key] = "latest"
+
+    esm = component._render_esm(compiled=False)
+    if issubclass(component, ReactComponent):
+        esm += '\nimport * as React from "react"\nimport { createRoot } from "react-dom/client"\nexport default {render, React, createRoot}'
+
+    js_code, packages = extract_packages(esm)
+    dependencies.update(packages)
+
+    # Create package.json structure
+    package_json = {
+        "name": camel_to_kebab(component.__name__),
+        "dependencies": dependencies
+    }
+    return js_code, package_json
+
+
+def compile_component(component: type[ReactiveESM], minify: bool = True):
+    """
+    Compiles and bundles the ESM code of a ReactiveESM component and
+    writes the compiled JS file to a file adjacent to the module
+    the component was defined in.
+    """
+    name = camel_to_kebab(component.__name__)
+    ext = 'jsx' if issubclass(component, ReactComponent) else 'js'
+    if hasattr(component, '__path__'):
+        out_path = component.__path__
+    else:
+        out_path = pathlib.Path(inspect.getfile(component)).parent
+    out = (pathlib.Path(out_path) / f'{name}.compiled.js').absolute()
+
+    code, package_json = extract_dependencies(component)
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+        package_json_path = os.path.join(temp_dir, 'package.json')
+        with open(package_json_path, 'w') as package_json_file:
+            json.dump(package_json, package_json_file, indent=2)
+
+        index_js_path = os.path.join(temp_dir, f'index.{ext}')
+        with open(index_js_path, 'w') as index_js_file:
+            index_js_file.write(code)
+
+        os.chdir(temp_dir)
+
+        install_cmd = ['npm', 'install']
```

Comment:
Do a check here. Maybe you can reuse what is already in `bokeh build`.

Reply:
Bokeh has a far more complex build chain and asks you to run `npm install` yourself.